src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -275,59 +275,107 @@ int IComparable.CompareTo(Object obj)
         // return an array of one byte whose element is 0x00.
         public byte[] ToByteArray()
         {
-            if (_bits == null && _sign == 0)
+            int sign = _sign;
+            if (sign == 0)
                 return new byte[] { 0 };
 
-            // We could probably make this more efficient by eliminating one of the passes.
-            // The current code does one pass for uint array -> byte array conversion,
-            // and then another pass to remove unneeded bytes at the top of the array.
-            uint[] dwords;
             byte highByte;
-
-            if (_bits == null)
+            int nonZeroDwordIndex = 0;
+            uint highDword;
+            uint[] bits = _bits;
+            if (bits == null)
             {
-                dwords = new uint[] { (uint)_sign };
-                highByte = (byte)((_sign < 0) ? 0xff : 0x00);
+                highByte = (byte)((sign < 0) ? 0xff : 0x00);
+                highDword = unchecked((uint)sign);
             }
-            else if (_sign == -1)
+            else if (sign == -1)
             {
-                dwords = (uint[])_bits.Clone();
-                NumericsHelpers.DangerousMakeTwosComplement(dwords);  // mutates dwords
                 highByte = 0xff;
+
+                // If sign is -1, we will need to two's complement bits.
+                // Previously this was accomplished via NumericsHelpers.DangerousMakeTwosComplement(),
+                // however, we can do the two's complement on the stack so as to avoid
+                // creating a temporary copy of bits just to hold the two's complement.
+                // One special case in DangerousMakeTwosComplement() is that if the array
+                // is all zeros, then it would allocate a new array with the high-order
+                // uint set to 1 (for the carry). In our usage, we will not hit this case
+                // because a bits array of all zeros would represent 0, and this case
+                // would be encoded as _bits = null and _sign = 0.
+                Debug.Assert(bits.Length > 0);
+                Debug.Assert(bits[bits.Length - 1] != 0);
+                while (bits[nonZeroDwordIndex] == 0U)
+                {
+                    nonZeroDwordIndex++;
+                }
+
+                highDword = ~bits[bits.Length - 1];
+                if (bits.Length - 1 == nonZeroDwordIndex)
+                    highDword += 1U;
             }
             else
             {
-                dwords = _bits;
+                Debug.Assert(sign == 1);
                 highByte = 0x00;
+                highDword = bits[bits.Length - 1];
+            }
+
+            byte msb;
+            int msbIndex;
+            if ((msb = (byte)((highDword & 0xff000000U) >> 24)) != highByte)
```

Comment:
The binary ands shouldn't be necessary, the shift/cast remove the other bits anyway.